controller: Proactively disconnect node based on heartbeat

dofmind opened the pull request:
This PR is an update of https://github.com/eclipse-bluechi/bluechi/pull/870

The main difference is that the code is simplified by using microseconds in last_seen instead of the struct timespec.
It updates last_seen in node_method_register() to prevent the node disconnecting right now when last_seen is 0 or too old.
Additionally, it handles disconnected nodes by calling node_disconnected() instead of controller_remove_node(). Otherwise an error will occur.
```
bluechi-controller: ../git/src/controller/node.c:123: unit_subscriptions_clear: Assertion `LIST_IS_EMPTY(usubs->subs)' failed.
```

Fixes: https://github.com/eclipse-bluechi/bluechi/issues/857

ueno committed:
Fix unused return value of sd_bus_match_signal

Signed-off-by: Daiki Ueno <dueno@redhat.com>

42dot-joonyoung-shim committed:
controller: Proactively disconnect node based on heartbeat

This adds a couple of new options to controller: HeartbeatInterval and
NodeHeartbeatThreshold, which can be used to actively disconnect node
based on the last sent heartbeat.  The controller periodically checks
the last seen timestamp of nodes, and if it was sent before
NodeHeartbeatThreshold, the controller treats it as disconnected.

Signed-off-by: Daiki Ueno <dueno@redhat.com>
Signed-off-by: Joonyoung Shim <joonyoung.shim@42dot.ai>

engelmi commented on doc/man/bluechi-controller.conf.5.md:
```diff
@@ -27,6 +27,14 @@ A comma separated list of unique bluechi-agent names. It's mandatory to set the
 
+#### **HeartbeatInterval** (long)
+
+The interval to periodically check node's connectivity based on heartbeat signals sent to bluechi, in milliseconds.
```
I think it would be good to mention here that a value of `0` disables the periodic check:
```suggestion
The interval to periodically check node's connectivity based on heartbeat signals sent to bluechi, in milliseconds. A value of 0 disables it. 
```

Not part of this PR, but this should be added to the description of the [Agents Heartbeat](https://github.com/eclipse-bluechi/bluechi/blob/main/config/agent/agent.conf#L29) as well.

dofmind replied:
done

engelmi replied:
Thanks!

engelmi commented on doc/man/bluechi-controller.conf.5.md:
```diff
@@ -27,6 +27,14 @@ A comma separated list of unique bluechi-agent names. It's mandatory to set the
+
+#### **NodeHeartbeatThreshold** (long)
+
+The threshold in milliseconds to determine whether a node is disconnected. If the node's last heartbeat signal was received before this threshold, bluechi treats that the node has been disconnected.
```
```suggestion
The threshold in milliseconds to determine whether a node is disconnected. If the node's last heartbeat signal was received before this threshold, bluechi assumes that the node is down or the connection was cut off and performs a disconnect.
```

dofmind replied:
done

engelmi commented on src/controller/controller.c:
```diff
@@ -444,6 +481,82 @@ static bool controller_setup_node_connection_handler(Controller *controller) {
+
+                diff = now - node->last_seen;
+                if (diff > (uint64_t) controller->heartbeat_threshold_msec * USEC_PER_MSEC) {
+                        node_disconnect(node);
```
It would be good to have an info log here to provide a bit of information why a node gets disconnected:
```suggestion
                       bc_log_infof("Did not receive heartbeat from node '%s' since '%d'ms. Disconnecting it...", node->name, controller->heartbeat_threshold_msec);
                        node_disconnect(node);
```

dofmind replied:
done

engelmi commented on src/controller/controller.c:
```diff
@@ -444,6 +481,82 @@ static bool controller_setup_node_connection_handler(Controller *controller) {
+
+                now = get_time_micros();
+                if (now == 0) {
+                        return -EINVAL;
```
Probably worth adding a log here:
```suggestion
                       bc_log_error("Failed to get the time");
                        return -EINVAL;
```

Also, the `return`ing here will result in `controller_reset_heartbeat_timer` not being called - so this would completely stop controller heartbeat. Is this desired?

dofmind replied:
Okay, let's handle this with a continue statement rather than a return statement.

engelmi commented on src/controller/controller.c:
```diff
@@ -444,6 +481,82 @@ static bool controller_setup_node_connection_handler(Controller *controller) {
+
+        if (controller->heartbeat_interval_msec <= 0) {
+                bc_log_warnf("Heartbeat disabled since interval is '%d', ",
+                             controller->heartbeat_interval_msec);
```
```suggestion
                bc_log_warnf("Heartbeat disabled since configured interval '%d' is <=0",
                             controller->heartbeat_interval_msec);
```

Again, not related to the PR, but could you change this as well [for the agent here](https://github.com/eclipse-bluechi/bluechi/blob/main/src/agent/agent.c#L190).

dofmind replied:
done

engelmi commented on config/controller/controller.conf:
```diff
@@ -16,6 +16,14 @@
 #ControllerPort=842
 
+#
+# Defines the interval between two heartbeat signals sent to bluechi in milliseconds.
```
```suggestion
# Defines the interval for which the bluechi-controller periodically checks the connectivity of all nodes based on the received heartbeat signals sent to bluechi-agent to bluechi-controller, in milliseconds.
```

dofmind replied:
done

engelmi commented on config/controller/controller.conf:
```diff
@@ -16,6 +16,14 @@
+#HeartbeatInterval=2000
+
+#
+# Defines the threshold to actively disconnect nodes based on the time difference from the last heartbeat signal.
```
```suggestion
# Defines the threshold to actively disconnect nodes based on the time difference from the last received heartbeat signal from the respective bluechi-agent. In milliseconds.
```

dofmind replied:
done

engelmi commented on config/controller/controller.conf:
```diff
@@ -16,6 +16,14 @@
 
+#
+# Defines the interval between two heartbeat signals sent to bluechi in milliseconds.
+#HeartbeatInterval=2000
```
The values here should reflect the default ones. 

For `HeartbeatInterval` I think a default value of `0` would be good to disable this feature by default. 
The default value of `6000` for the `NodeHeartbeatThreshold` seems reasonable to me. 

Also, we decided to hard-code the defaults in and set these before parsing any configuration files. Therefore, please add 
- the default values also in the [protocol.h](https://github.com/eclipse-bluechi/bluechi/blob/main/src/libbluechi/common/protocol.h#L29) next to the other defaults. 
- setting these hard-coded values in the [cfg_controller_def_conf](https://github.com/eclipse-bluechi/bluechi/blob/main/src/libbluechi/common/cfg.c#L490-L511)

dofmind replied:
done

engelmi reviewed (commented):
Thanks for your contribution! @dofmind 

Overall the PR looks good. I added a few minor comments, mostly about logging. I tested your changes manually (also via valgrind) and it works well. Nonetheless, please add two integration tests:
- checking that `HeartbeatInteral=0` disables it 
(e.g. by confirming the agent stays connected after x sec when both, agent and controller heartbeat are disabled and controller threshold is y sec)
- checking that the agent gets disconnected when not the threshold is reached (e.g. by disabling the agent heartbeat)

42dot-joonyoung-shim committed:
Add integration tests for heartbeat

An integration test is to verify if default configuration of controller
disables periodic heartbeat of the controller, and the other is to
verify if the node gets disconnected when did not receive heartbeat
since threshold from node.

Signed-off-by: Joonyoung Shim <joonyoung.shim@42dot.ai>

engelmi reviewed (approved):
LGTM

Created an additional documentation issue for extending the readthedocs https://github.com/eclipse-bluechi/bluechi/issues/913
